libp2p/identity/identity.go

```diff
@@ -0,0 +1,134 @@
+package identity
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multiaddr"
+
+	"github.com/pokt-network/pocket/p2p/transport"
+	"github.com/pokt-network/pocket/p2p/types"
+	poktCrypto "github.com/pokt-network/pocket/shared/crypto"
+)
+
+var (
+	ErrIdentity = types.NewErrFactory("")
+)
+
+// PoktPeerFromStream builds a pokt peer from a libp2p stream.
+// (NOTE: excludes `ServiceURL`)
+func PoktPeerFromStream(stream network.Stream) (*types.NetworkPeer, error) {
+	remotePubKeyBytes, err := stream.Conn().RemotePublicKey().Raw()
+	// NB: abort handling this stream.
+	if err != nil {
+		return nil, err
+	}
+	poktPubKey, err := poktCrypto.NewPublicKeyFromBytes(remotePubKeyBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.NetworkPeer{
+		Dialer:    transport.NewLibP2PTransport(stream),
+		PublicKey: poktPubKey,
+		// NB: pokt analogue of libp2p peer.ID
+		Address:    poktPubKey.Address(),
+		ServiceUrl: stream.Conn().RemoteMultiaddr().String(),
+	}, nil
+}
+
+// PubKeyFromPoktPeer retrieves the libp2p compatible public key from a pocket peer.
+func PubKeyFromPoktPeer(poktPeer *types.NetworkPeer) (crypto.PubKey, error) {
+	pubKey, err := crypto.UnmarshalEd25519PublicKey(poktPeer.PublicKey.Bytes())
+	if err != nil {
+		return nil, ErrIdentity(fmt.Sprintf(
+			"unmarshalling peer ed25519 public key, pokt address: %s", poktPeer.Address,
+		), err)
+	}
+
+	return pubKey, nil
+}
+
+// PeerAddrInfoFromPoktPeer builds a libp2p AddrInfo which maps to the passed pocket peer.
+func PeerAddrInfoFromPoktPeer(poktPeer *types.NetworkPeer) (peer.AddrInfo, error) {
+	pubKey, err := PubKeyFromPoktPeer(poktPeer)
+	if err != nil {
+		return peer.AddrInfo{}, err
+	}
+
+	peerID, err := peer.IDFromPublicKey(pubKey)
+	if err != nil {
+		return peer.AddrInfo{}, ErrIdentity(fmt.Sprintf(
+			"retrieving ID from peer public key, pokt address: %s", poktPeer.Address,
+		), err)
+	}
+
+	peerMultiaddr, err := multiaddr.NewMultiaddr(poktPeer.ServiceUrl)
+	// NB: early return if we already have a multiaddr.
+	if err == nil {
+		return peer.AddrInfo{
+			ID: peerID,
+			Addrs: []multiaddr.Multiaddr{
+				peerMultiaddr,
+			},
+		}, nil
+	}
+
+	peerMultiaddr, err = PeerMultiAddrFromServiceURL(poktPeer.ServiceUrl)
+	if err != nil {
+		return peer.AddrInfo{}, err
+	}
+
+	return peer.AddrInfo{
+		ID: peerID,
+		Addrs: []multiaddr.Multiaddr{
+			peerMultiaddr,
+		},
+	}, nil
+}
+
+// TECHDEBT: this probably belongs somewhere else, it's more of a networking helper.
+func PeerMultiAddrFromServiceURL(serviceURL string) (multiaddr.Multiaddr, error) {
+	// NB: hard-code a scheme for URL parsing to work.
```

Comment:
Do we always expect `scheme` to come from LibP2P service URLs?

If so: Move to const and add a comment
If not: add answer below

Reply:
I'm not sure if I followed the question exactly. My comment might not have been phrased clearly. To elaborate: we need to append something to fill the "scheme" space in the URL in order for the URL parser to be happy. The only examples I've looked at are the addresses in the genesis, none of which includes a scheme. I considered that we could use scheme to signal the expected transport to remote peers as it's currently only exposed to each node locally via their config (connection type). I figured that we would consider switching to multiaddrs, which would eliminate the need for that discussion.